File: solutions/go/rail-fence-cipher/1/rail_fence_cipher.go
// For my Shree DR.MDD
package railfence

import "strings"

func railPattern(length int, numRails int) []int {
	posList := make([]int, length)
	for i, step, idx := 0, 1, 0; i < length; i++ {
		posList[i] = idx
		switch {
		case idx == 0:
			step = 1
		case idx == numRails-1:
			step = -1
		}
		idx += step
	}
	return posList
}

func railCounts(length int, numRails int) []int {
	counts := make([]int, numRails)
	for _, p := range railPattern(length, numRails) {
		counts[p]++
	}
	return counts
}

func splitMessageByRails(msg string, numRails int) []string {
	counts := railCounts(len(msg), numRails)
	result := make([]string, numRails)

	start, end := 0, 0
	for i, c := range counts {
		end = start + c
		result[i] = msg[start:end]
		start = end
	}
	return result
}

func Encode(msg string, numRails int) string {
	seq := railPattern(len(msg), numRails)
	rails := make([]string, numRails)
	for i, ch := range msg {
		idx := seq[i]
		rails[idx] += string(ch)
	}
	return strings.Join(rails, "")
}

func Decode(msg string, numRails int) string {
	decoded := ""

	seq := railPattern(len(msg), numRails)
	rails := splitMessageByRails(msg, numRails)
	for _, idx := range seq {
		decoded += rails[idx][:1]
		rails[idx] = rails[idx][1:]
	}
	return decoded
}

File: solutions/lua/rail-fence-cipher/1/rail-fence-cipher.lua
-- For my Shree DR.MDD
local function merge_tables(tables)
  local merged = {}
  for _, tbl in ipairs(tables) do
    for _, item in ipairs(tbl) do
      table.insert(merged, item)
    end
  end
  return merged
end

local function rail_positions(len, rails)
  local tracks = {}
  local idx = 1
  local step = 1

  for i = 1, rails do tracks[i] = {} end

  for i = 1, len do
    table.insert(tracks[idx], i)
    idx = idx + step
    if idx == rails then step = -1 end
    if idx == 1 then step = 1 end
  end

  return ipairs(merge_tables(tracks))
end

local function encode(plain, rails)
  local cipher = {}
  for i, pos in rail_positions(#plain, rails) do
    cipher[i] = plain:sub(pos, pos)
  end
  return table.concat(cipher)
end

local function decode(cipher, rails)
  local plain = {}
  for i, pos in rail_positions(#cipher, rails) do
    plain[pos] = cipher:sub(i, i)
  end
  return table.concat(plain)
end

return { encode = encode, decode = decode }
